x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetReader.java

```diff
@@ -85,31 +84,21 @@ private static int computeNumDocs(LeafReader reader, BitSet roleQueryBits) {
     /**
      * Like {@link #computeNumDocs} but caches results.
      */
-    private static int getNumDocs(LeafReader reader, Query roleQuery, BitSet roleQueryBits) throws IOException, ExecutionException {
+    private static int getNumDocs(LeafReader reader, Query roleQuery, BitSet roleQueryBits) throws ExecutionException {
         IndexReader.CacheHelper cacheHelper = reader.getReaderCacheHelper(); // this one takes deletes into account
         if (cacheHelper == null) {
             return computeNumDocs(reader, roleQueryBits);
         }
-        final boolean[] added = new boolean[] { false };
-        Cache<Query, Integer> perReaderCache = NUM_DOCS_CACHE.computeIfAbsent(cacheHelper.getKey(), key -> {
-            added[0] = true;
-            return CacheBuilder.<Query, Integer>builder()
+        return NUM_DOCS_CACHE.computeIfAbsent(cacheHelper.getKey(), key -> {
+            var res = CacheBuilder.<Query, Integer>builder()
                 // Not configurable, this limit only exists so that if a role query is updated
                 // then we won't risk OOME because of old role queries that are not used anymore
                 .setMaximumWeight(1000)
                 .weigher((k, v) -> 1) // just count
                 .build();
-        });
-        if (added[0]) {
-            IndexReader.ClosedListener closedListener = NUM_DOCS_CACHE::remove;
-            try {
-                cacheHelper.addClosedListener(closedListener);
-            } catch (AlreadyClosedException e) {
```

Comment:
Why aren't we still handle the `AlreadyClosedException` case?

Reply:
That's because it would now happen in the `computeIfAbsent` callback which would fail adding to the map, hence we don't need to clean up after it :)

Reply:
Ah, yes, I missed that `closedListener.onClose(cacheHelper.getKey())` effectively removes the mapping.